### DISCUSSION TITLE
Added: LDM Notes for June 7th, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#4811] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-06-07.md

## Agenda

1. Runtime checks for parameterless struct constructors
2. List patterns
    a. Exhaustiveness
    b. Length pattern feedback

**Comment by erikhermansson79** [comment]
My suggestion for list patterns is to always have both brackets and curly braces, even if they are empty. So the empty list would be `[]{}` or `[0]{}`.

**Comment by MgSam** [comment]
I've read the list pattern proposal more than once and the design notes and I still have absolutely no idea what 
``` C#
_ = list switch
{
    { .., >= 0 } => 1,
    { < 0 } => 2,
    { Count: <= 0 or > 1 } => 3,
};
```
or 
``` C#
_ = list switch
{
    { .., >= 0 } => 1,
    { ..{ .., < 0 } } => 3,
};
```
are supposed to do. I'm really am not a fan of features that ditch clarity in favor of brevity.

**Reply by bernd5** [reply]
I try to explain the pattern AFAIK:
- `{ .., >= 0 }` checks if the collection/ list starts with some elements and ends with an element for witch `>= 0` is true
- `{ < 0 }` has a single element which satisfies the pattern `< 0`
- `{ Count: <= 0 or > 1 }` is not a list but a property pattern and checks if `Count` is less or equal to zero or greater 1 - or logical if the first element is not equal to 1
- `{ .., >= 0 }` starts with some elements and the final element should satisfy `>= 0`
- `{ ..{ .., < 0 } }` checks if the last element is `< 0` - but in a fairly complicated way

**Reply by CyrusNajmabadi** [reply]
```c#
_ = list switch
{
    { .., >= 0 } => 1,
    { < 0 } => 2,
    { Count: <= 0 or > 1 } => 3,
};
```

This is testing three cases.  Let's break them down:

```{ .., >= 0 }```

This is: Can start with any number of elements, with a trailing element that must be greater or equal to 0.

```{ < 0 }```

This is: Must have exactly one element in it.  It must be less than 0.

```{ Count: <= 0 or > 1 }```

This is better written as:

```{ Count: != 1 }```

As is probably pretty clear for this, this is any list whose count is not equal to 1.

**Reply by CyrusNajmabadi** [reply]
The second switch isn't meant as an example of what you should write.  It's meant for us to explore what it would mean if someone wrote it.  It's meant so we can determine what invariatns should/shouldn't hold true.  For completeness though:

```{ ..{ .., < 0 } }```

This means you have have any number of elements in your list.  However, those elements themselves can have any number of elements, but must end with an element that is less than 0.  Intuitively, that means that it should be the same as `{ .., < 0 }`.  However, nothing in the spec made that claim.  This meeting was about us defining the rules such that from the language perspective we *do* consider those as having the same meanign for the purposes of our own analysis of your code.  It is not a recommendation that anyone actually write this.

**Reply by bernd5** [reply]
@CyrusNajmabadi Good comment - I was just surprised about `{ Count: != 1 }` - AFAIK this is currently invalid syntax. You would have to write `{Count: not 1}`.
Is that syntax planed for future versions of C#?

**Reply by jnm2** [reply]
> I'm really am not a fan of features that ditch clarity in favor of brevity.

As counterpoint, it's painful to me to have to exit out of a nested pattern and capture a new variable and then index into it as a separate pattern. Nested `{ Arguments: { _, { X: Y } secondArgument, .. } }` is a lot nicer than having to declare a variable for the arguments list and then index into it as a separate pattern.

**Reply by 333fred** [reply]
> @CyrusNajmabadi Good comment - I was just surprised about `{ Count: != 1 }` - AFAIK this is currently invalid syntax. You would have to write `{Count: not 1}`.
> Is that syntax planed for future versions of C#?

No, it is not.

**Reply by MgSam** [reply]
Thanks all for the explanations. 

It feels a little weird to me to have yet another way to refer to items from the end of a list, given that `[^1]` was just added not all that long ago. That said, I can't really think of a good way to integrate that syntax with what this syntax is trying to accomplish.

**Comment by amis92** [comment]
What if list patterns, list deconstructors, and a new list initialization pattern is added as well for symmetry - this time maybe taking immutable/init-only lists as well?

It'd require a new list syntax overall. Maybe `[| a, b |]` as F# has?

**Reply by amis92** [reply]
@CyrusNajmabadi @333fred Any chance you took a look? I know it's a huge change and very... bold proposal, but since square brackets have been associated with array length, ranges, and indexes, curly braces with object initializers/patterns, maybe a separate, new syntax could be introduced for a "new era" and cover all operations intuitively?

**Reply by CyrusNajmabadi** [reply]
Yes. On Monday we approved a new syntax for list patterns

**Reply by 333fred** [reply]
My notes for the meeting will be up later today, but it's not new syntax. It's just square brackets.

**Reply by HaloFour** [reply]
See: https://github.com/dotnet/csharplang/discussions/4830#discussioncomment-878072

**Reply by amis92** [reply]
Thanks all for the responses. [Cyrus's comment](https://github.com/dotnet/csharplang/discussions/4830#discussioncomment-879885) about possible new list construction syntax gives me hope though that not all is lost for immutable collections :)

**Reply by HaloFour** [reply]
I don't believe the conversation waded into immutable collections, only that the compiler would support a new form of collection initializer with target typing, e.g. `List<int> list = [ 1, 2, 3, 4, 5 ];`.  Perhaps a new discussion should be started for the list initialization syntax and these kinds of questions can be put on the table.

**Reply by CyrusNajmabadi** [reply]
A new discussion woudl be good (if we don't have one already).  I can state emphatically though that i'm only in support of a list-literal proposal that supports immutable collections :)

**Reply by HaloFour** [reply]
There's #414 which focuses on dictionary literals but also touches on list literals (enough so that future list literal proposals were closed as dupes referencing it).

IMO it's probably worthwhile to consider both list/dictionary literals and corresponding patterns at the same time.  And it's a great opportunity to review all of the cases where current collection/indexer initializer syntax isn't suitable due to being inefficient or expecting to mutate the current instance.
